### ISSUE TITLE
Upgrade to v18.4 failed, cannot install both kubelet-1.18.4-1.x86_64 and kubelet-1.18.4-0.x86_64

### ISSUE THREAD

**Issue by StefanSa** [kubernetes/kubernetes#92463]
**What happened**:
on the master `centos-8` there is the following error message when trying to update the packages.

``` 
Problem: cannot install both kubelet-1.18.4-1.x86_64 and kubelet-1.18.4-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.7.5-0.x86_64
  - cannot install the best update candidate for package kubelet-1.18.3-0.x86_64
```
We don't see this problem in the nodes `centos-7`
Likewise, we did not see this error in the master `centtos-8` in the last update to v18.3.

**How to reproduce it (as minimally and precisely as possible)**:
yum update / dnf update

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:  vmware
- OS (e.g: `cat /etc/os-release`): centos-8

**Comment by StefanSa** [comment]
This helps me as a short-term solution.
`yum update --skip-broken --nobest`

but does not fix the actual packet conflict.

```
yum update
 Problem: cannot install both kubelet-1.18.4-0.x86_64 and kubelet-1.18.4-1.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.6-0.x86_64
  - cannot install the best update candidate for package kubelet-1.18.4-1.x86_64
```

**Comment by justaugustus** [comment]
@StefanSa -- What happens when you completely remove `kubelet` and `kubernetes-cni` and then attempt to install one of the following kubelets:
- kubelet-1.18.5-0
- kubelet-1.18.4-1

(Related to https://github.com/kubernetes/kubernetes/issues/92242.)
/assign
/area release-eng
cc: @kubernetes/release-engineering

**Comment by StefanSa** [comment]
@justaugustus
i  remove `kubelet` and `kubernetes-cni`and installed it again successfully with:
`yum install kubelet-1.18.5 kubeadm-1.18.5`, however, a `yum update` still fails.

``` 
yum update
Problem: cannot install both kubelet-1.18.4-0.x86_64 and kubelet-1.18.5-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.6-0.x86_64
  - cannot install the best update candidate for package kubelet-1.18.5-0.x86_64
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken'
to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```
As already mentioned, we don't see that on our nodes that are based on centos 7, currently we only see this problem with the master that is based on centos 8.

**Comment by mrbobbytables** [comment]
There does still seem to be an issue on centos8 I had the same issue when testing with 1.17.8, 1.18.1, 1.18.3, and 1.18.5
```
root@a77d16cf886f /]# yum info kubelet
Failed to set locale, defaulting to C.UTF-8
Last metadata expiration check: 0:06:28 ago on Thu Jul 30 12:22:37 2020.
Installed Packages
Name         : kubelet
Version      : 1.18.5
Release      : 0
Architecture : x86_64
Size         : 108 M
Source       : kubelet-1.18.5-0.src.rpm
Repository   : @System
From repo    : kubernetes
Summary      : Container cluster management
URL          : https://kubernetes.io
License      : ASL 2.0
Description  : The node agent of Kubernetes, the container cluster manager.

Available Packages
Name         : kubelet
Version      : 1.18.6
Release      : 0
Architecture : x86_64
Size         : 21 M
Source       : kubelet-1.18.6-0.src.rpm
Repository   : kubernetes
Summary      : Container cluster management
URL          : https://kubernetes.io
License      : ASL 2.0
Description  : The node agent of Kubernetes, the container cluster manager.

[root@a77d16cf886f /]# yum update
Failed to set locale, defaulting to C.UTF-8
Last metadata expiration check: 0:06:33 ago on Thu Jul 30 12:22:37 2020.
Error:
 Problem: cannot install both kubelet-1.18.6-0.x86_64 and kubelet-1.18.4-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.6-0.x86_64
  - cannot install the best update candidate for package kubelet-1.18.5-0.x86_64
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```

**Comment by huataihuang** [comment]
I still had same issue with 1.18.6 upgade to 1.19.0

```
#dnf update
...
Error:
 Problem: cannot install both kubelet-1.19.0-0.x86_64 and kubelet-1.18.4-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.6-0.x86_64
  - cannot install the best update candidate for package kubelet-1.18.6-0.x86_64
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```

```
#dnf info kubelet
Last metadata expiration check: 0:39:13 ago on Wed 02 Sep 2020 09:18:41 AM CST.
Installed Packages
Name         : kubelet
Version      : 1.18.6
Release      : 0
Architecture : x86_64
Size         : 108 M
Source       : kubelet-1.18.6-0.src.rpm
Repository   : @System
From repo    : kubernetes
Summary      : Container cluster management
URL          : https://kubernetes.io
License      : ASL 2.0
Description  : The node agent of Kubernetes, the container cluster manager.

Available Packages
Name         : kubelet
Version      : 1.19.0
Release      : 0
Architecture : x86_64
Size         : 19 M
Source       : kubelet-1.19.0-0.src.rpm
Repository   : kubernetes
Summary      : Container cluster management
URL          : https://kubernetes.io
License      : ASL 2.0
Description  : The node agent of Kubernetes, the container cluster manager.
```

**Comment by huataihuang** [comment]
I add follow config to `/etc/yum.conf` to exclude conflict kubelet

```
exclude=kubelet-1.18.4-0 kubelet-1.17.7-0 kubelet-1.16.11-0
```

then use `dnf upgrade` to upgrade kubelet 1.19.0 without problem.

I hope it can help you.

**Comment by laundry-96** [comment]
@huataihuang What repository url do you have? I've been using https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64 (which is what the kubernetes website tells me to use) and it doesn't seem to have kubelet-1.19 in it...

**Comment by huataihuang** [comment]
> @huataihuang What repository url do you have? I've been using https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64 (which is what the kubernetes website tells me to use) and it doesn't seem to have kubelet-1.19 in it...

https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64

here is my `/etc/yum.repos.d/kubernetes.repo`

```
[kubernetes]
name=Kubernetes
baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
```

**Comment by justaugustus** [comment]
/assign @saschagrunert 
/unassign

Sascha -- can you take a look or assign to someone on @kubernetes/release-engineering to investigate?

The following issues are similar classes of issues for apt/yum:
- https://github.com/kubernetes/kubernetes/issues/92698
- https://github.com/kubernetes/kubernetes/issues/92587
- https://github.com/kubernetes/kubernetes/issues/92586
- https://github.com/kubernetes/kubernetes/issues/92463

**Comment by saschagrunert** [comment]
/help
Not sure if we can solve this without re-building those packages.

**Comment by Mikhail2577** [comment]
Hi everyone, seems like I can't install kubeadm or kubelet on my fresh install of CentOS7. 

sudo yum install -y kubelet kubeadm kubectl --disableexcludes=kubernetes

Loaded plugins: fastestmirror, product-id, search-disabled-repos, subscription-manager

This system is not registered with an entitlement server. You can use subscription-manager to register.

Loading mirror speeds from cached hostfile
 * base: ftp.belnet.be
 * extras: ftp.belnet.be
 * updates: ftp.belnet.be
No package kubelet available.
No package kubeadm available.
No package kubectl available.
Error: Nothing to do

**Comment by CyshallChan** [comment]
I use centos 8, same issue.

**Comment by likan999** [comment]
Just curious, why doesn't Kubernetes support CentOS 8? We still need to use el7 repos on CentOS 8.

**Comment by yabb85** [comment]
I use centos 8, same issue.

**Comment by boogiefromzk** [comment]
Same on centos 8

**Comment by SnailDove** [comment]
there also is the same on my CentOS 8 system

**Comment by boogiefromzk** [comment]
Still actual

**Comment by cpanato** [comment]
This happened for 1.18 right? this is not supposed to occur on 1.19+ releases, I guess.

1.18 reach the end of life and we are not supporting this version anymore.
I guess we can close this issue.

comments? @saschagrunert @justaugustus @hasheddan

**Comment by vrusinov** [comment]
Conflicts similar to above still happen on CentOS 8 with 1.21 rpms installed. `exclude=` in `yum.conf` mitigates it, but it would be nice to clean this up somehow.

**Comment by TheFrisianClause** [comment]
Well I am trying to automate this process with Ansible as well on RockyLinux 8 but I seem to get this error to. This is the debug:

fatal: [kube-worker-02]: FAILED! => {"changed": false, "failures": [], "msg": "Depsolve Error occured: \n Problem: cannot install both kubelet-1.18.4-0.x86_64 and kubelet-1.22.0-0.x86_64\n  - cannot install the best update candidate for package kubernetes-cni-0.8.7-0.x86_64\n  - cannot install the best update candidate for package kubelet-1.22.0-0.x86_64", "rc": 1, "results": []}

I also tried 'Allowerasing: yes' but this seem not to resolve the problem unfortunately...

I have the following kubernetes repo configured (according to the kubernetes documentation)

[kubernetes]
name=Kubernetes
baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-$basearch
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
exclude = kubelet kubeadm kubectl

Is there a reason why this happens and if so is there some kind of fix on the way to resolve this? As it seems to be going on for quite a while now.

**Comment by Michael-M-Zhang** [comment]
Same issue on my centos system.
sudo yum update
Repository extras is listed more than once in the configuration
Last metadata expiration check: 2:48:25 ago on Wed 15 Sep 2021 07:04:01 PM CST.
Error:
 Problem: cannot install both kubelet-1.18.4-0.x86_64 and kubelet-1.22.1-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.7-0.x86_64
  - cannot install the best update candidate for package kubelet-1.22.1-0.x86_64
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

**Comment by vrusinov** [comment]
/remove-lifecycle stale

This is still happening. In general kubernetes RPM repository is not in the best shape.

**Comment by fusionneur** [comment]
still heppening:
 Problem: cannot install both kubelet-1.18.4-0.x86_64 and kubelet-1.23.2-0.x86_64
  - cannot install the best update candidate for package kubernetes-cni-0.8.7-0.x86_64
  - cannot install the best update candidate for package kubelet-1.23.2-0.x86_64

**Comment by ViliusS** [comment]
So almost 2 years later and dozen of duplicate issues filled against this bug, we still have the same issue :(

**Comment by fusionneur** [comment]
> I add follow config to `/etc/yum.conf` to exclude conflict kubelet
> 
> ```
> exclude=kubelet-1.18.4-0 kubelet-1.17.7-0 kubelet-1.16.11-0
> ```
> 
> then use `dnf upgrade` to upgrade kubelet 1.19.0 without problem.
> 
> I hope it can help you.

This is the workaround. I'll repost this, because so that it is not lost after 2 years of comments. At least to help others bumping into this issue..

**Comment by aimcod** [comment]
> I add follow config to `/etc/yum.conf` to exclude conflict kubelet
> 
> ```
> exclude=kubelet-1.18.4-0 kubelet-1.17.7-0 kubelet-1.16.11-0
> ```
> 
> then use `dnf upgrade` to upgrade kubelet 1.19.0 without problem.
> 
> I hope it can help you.

It did help, quite a lot. Thanks for this.

What I ended up doing, was to `exclude=kubelet-1.1*`

I am adding this to my ansible kubernetes deployment playbooks. Sweet and short.

Thanks a ton!

**Comment by nickson** [comment]
I have the same issue with Alma Linux 9.1:
`dnf update
...
 Problem: cannot install both kubelet-1.26.1-0.x86_64 and kubelet-1.18.4-0.x86_64`

The workaround with the exclude fixes it:
`
tail -1 /etc/yum.repos.d/kubernetes.repo
exclude = kubelet-1.18.* kubelet-1.17.* kubelet-1.16.*
`

**Comment by delfws** [comment]
same queastion in CentOS 9 Stream

```bash
root in ~
≥ uname -a                                                                                                         9:01
Linux master 5.14.0-307.el9.x86_64 #1 SMP PREEMPT_DYNAMIC Wed May 3 06:16:28 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

root in ~
≥ yum upgrade                                                                                                      8:44
上次元数据过期检查：0:06:42 前，执行于 2023年05月10日 星期三 08时38分13秒。
错误：
 问题: 无法同时安装 kubelet-1.18.4-0.x86_64 和 kubelet-1.27.1-0.x86_64
  - 无法为软件包安装最佳更新候选 kubernetes-cni-1.2.0-0.x86_64
  - 无法为软件包安装最佳更新候选 kubelet-1.27.1-0.x86_64
(尝试在命令行中添加 '--allowerasing' 来替换冲突的软件包 或 '--skip-broken' 来跳过无法安装的软件包 或 '--nobest' 来不只使用软件包的最佳候选)
```

**Comment by aojea** [comment]
> 1.18 reach the end of life and we are not supporting this version anymore.
> I guess we can close this issue.

last supported version is 1.24 https://kubernetes.io/releases/

There is no support for upgrading since 1.18 to any of the supported versions https://kubernetes.io/releases/version-skew-policy/#supported-component-upgrade-order

/close

**Comment by shawnz** [comment]
@aojea I think there's a misunderstanding. This bug applies to the latest versions even though the error message presented incorrectly mentions 1.18. The problem is that running "yum upgrade" will attempt to install 1.18 even if you are using a much newer version already.

**Comment by vrusinov** [comment]
Yes, the bug is not related to whether 1.18 is supported or not. It still happens when trying to update to latest Kubernetes. Please reopen.

PS: actually removing unsupported Kubernetes versions from the main yum repo may help. Another option is to have separate repositories for each of the major releases.

/reopen

**Comment by ViliusS** [comment]
@aojea please reopen. The title of this issue is incorrect, this happens when trying to update to any latest version. Even from 1.25.x to 1.26.x.

**Comment by aojea** [comment]
> @aojea please reopen. The title of this issue is incorrect, this happens when trying to update to any latest version. Even from 1.25.x to 1.26.x.

all the snippets in the code are with 1.18, can you paste a snippet with those versions ?

**Comment by vrusinov** [comment]
https://github.com/kubernetes/kubernetes/issues/92463#issuecomment-1442659517 is trying to install 1.26.

**Comment by aojea** [comment]
install 1.26 ... from 1.18

 anything that is not on a supported version (latest 1.24) and within the supported skew (n-2 versions respect the control-plane) is not supported https://github.com/kubernetes/kubernetes/issues/92463#issuecomment-1541763688

**Comment by shawnz** [comment]
@aojea No, the issue occurs no matter what the original and target versions are. No matter what version you are upgrading from or to, 1.18 will incorrectly get marked for installation. The issue does not require 1.18 to be the original or target version in the upgrade.

**Comment by aojea** [comment]
>  Even if you are upgrading from 1.25 to 1.26 for example, it will still mark 1.18 to be installed if you use "yum update".

lol, I didn't understand that , sorry 

/reopen

do we have 1.18 hardcoded somewhere?
/sig release

**Comment by aojea** [comment]
I think that the rpms sources are here https://github.com/kubernetes/release/tree/master/packages/rpm

/transfer kubernetes/release
